pkg/scheduler/preemption/preemption.go

```diff
@@ -361,20 +361,26 @@ func workloadFits(wlReq cache.FlavorResourceQuantities, cq *cache.ClusterQueue,
 				cohortResRequestable = cq.Cohort.RequestableResources[flvQuotas.Name]
 			}
 			for rName, rReq := range flvReq {
-				limit := flvQuotas.Resources[rName].Nominal
-
-				// No need to check whether cohort is nil here because when borrowingLimit
-				// is non nil, cohort is always non nil.
-				if allowBorrowing && flvQuotas.Resources[rName].BorrowingLimit != nil {
-					limit += *flvQuotas.Resources[rName].BorrowingLimit
-				}
-
-				if cqResUsage[rName]+rReq > limit {
-					return false
+				resource := flvQuotas.Resources[rName]
+
+				if cq.Cohort != nil {
+					if allowBorrowing {
+						// When resource.BorrowingLimit == nil there is no borrowing
+						// limit, so we can skip the check.
+						if resource.BorrowingLimit != nil {
+							if cqResUsage[rName]+rReq > resource.Nominal+*resource.BorrowingLimit {
+								return false
+							}
+						}
+					}
+					if cohortResUsage[rName]+rReq > cohortResRequestable[rName] {
+						return false
+					}
 				}
-
-				if cq.Cohort != nil && cohortResUsage[rName]+rReq > cohortResRequestable[rName] {
-					return false
+				if cq.Cohort == nil || !allowBorrowing {
```

Comment:
This check should probably go first:

```
if cohort == nil || !allowBorrowing {
  check against nominal
} else {
  limit := cohortRequestable
  if borrowingLimit != nil {
    limit = min(limit, nominal+borrowingLimit)
  }
  check against limit
}
```

Reply:
1. not that simple, because when `cohort != nil`, but regardless of `allowBorrowing` we need to check this (because allowBorrowing is CQ-scoped, but there might be other CQs borrowing already):
```golang
	if cohortResUsage[rName]+rReq > cohortResRequestable[rName] {
	   return false
	}
```
Yes, we could do something like 
```
  limit := cohortRequestable
  if borrowingLimit != nil {
    limit = min(limit, nominal+borrowingLimit)
  }
  ```
  but the point is that if `borrowingLimit == nil` then we don't do the check against the limit, which was source of the bug actually, so I wanted to make it step-by-step. 
  
  Still, I'm open to suggestions to improve readability, just pointing out it is not as simple as it may seem from the pseudocode.

Reply:
Just pushed an attempt to make it more readable sort of meeting in the middle with the suggestion. PTAL.